### PR TITLE
use jest for testing

### PR DESCRIPTION
**Pull request by justinwoo**
This adds Jest for testing AND generating coverage reports, keeps some of the testing setup the same, and mostly removes mocha, chokidar (jest has its own watcher), chai, and sinon (jest comes with its own mocking tools, and sinon was destroying the world in jest tests).

The warnings that come up when running this are the same as with the mocha tests.

For painless installation and testing, use `"react-native-lock": "justinwoo/react-native-lock#feature-customized-theme",` in package.json and run `yarn && jest` 😛 (fixed yarn bug but they still haven't released a new version) (but seriously, takes me 4 min to install this and 20s with yarn)

Relates to #121

### CODE REVIEW

**Comment by tino-junge** [comment]
@justinwoo With yarn v0.16.1 the installation is finally painless without changing the `react-native-lock` in the `package.json` ;) I can update the `yarn.lock` file for it.

File: support/rename.sh [hunk 1]
```diff
@@ -7,18 +7,32 @@ appRoot=`dirname $0`/..
 
 [ -z "${newAppName}" ] && echo 'Missing required parameter newAppName' && exit 1
 
-grep -rI 'PepperoniAppTemplate' --exclude='rename.sh' $appRoot/* | tr ':' ' ' | awk '{print $1}' | uniq | xargs -I{} sed -i.bak "s/PepperoniAppTemplate/${newAppName}/g" {}
-grep -rI 'pepperoniapptemplate' --exclude='rename.sh' $appRoot/* | tr ':' ' ' | awk '{print $1}' | uniq | xargs -I{} sed -i.bak "s/pepperoniapptemplate/${newLowerCaseName}/g" {}
-find . -name '*.bak' -exec rm {} \;
+# gather all modification targets
```
Comment (tino-junge) [inline support/rename.sh]: I guess that's not `jest` related and could've been done in a separate PR ;) I like adding the renamed files directly to git :) However, after running the new rename script I got the error in the simulator that the app hasn't been registered.

**Comment by jhabdas** [comment]
Happy to see this! Could Jest be used to test for component state mutation, to allow removal of Immutable in favor of POJOs?
